internal/crypto/src/time_stamp/verify.rs

```diff
@@ -0,0 +1,368 @@
+                                continue;
+                            }
+                        };
+
+                    let mut h = digest_algorithm.digester();
+                    if let Some(content) = &sd.content_info.content {
+                        h.update(&content.to_bytes());
+                    }
+
+                    let digest = h.finish();
+
+                    if signed_message_digest != digest.as_ref() {
+                        last_err = TimeStampError::InvalidData;
+                        continue;
+                    }
+                }
+
+                None => {
+                    last_err = TimeStampError::InvalidData;
+                    continue;
+                }
+            }
+        }
+
+        // Build CMS structure to verify.
+        let tbs = match signer_info.signed_attributes_digested_content() {
+            Ok(sdc) => match sdc {
+                Some(tbs) => tbs,
+                None => match &sd.content_info.content {
+                    Some(d) => d.to_bytes().to_vec(),
+                    None => {
+                        return Err(TimeStampError::DecodeError(
+                            "time stamp does not contain digested content".to_string(),
+                        ))
+                    }
+                },
+            },
+            Err(_) => {
+                last_err = TimeStampError::DecodeError(
+                    "time stamp does not contain digested content".to_string(),
+                );
+                continue;
+            }
+        };
+
+        let hash_alg = &signer_info.digest_algorithm.algorithm;
+        let sig_alg = &signer_info.signature_algorithm.algorithm;
+
+        // Grab signing certificate.
+        let sig_val = &signer_info.signature;
+        let mut signing_key_der = Vec::<u8>::new();
+        cert.tbs_certificate
+            .subject_public_key_info
+            .encode_ref()
+            .write_encoded(bcder::Mode::Der, &mut signing_key_der)?;
+
+        // Verify signature of time stamp signature.
+        #[cfg(feature = "openssl")]
+        validate_timestamp_sig(sig_alg, hash_alg, sig_val, &tbs, &signing_key_der)?;
+
+        #[cfg(not(feature = "openssl"))]
+        {
+            #[cfg(target_arch = "wasm32")]
+            {
+                if _sync {
+                    // IMPORTANT: The synchronous implementation of validate_timestamp_sync
+                    // on WASM is unable to support _some_ signature algorithms. The async path
+                    // should be used whenever possible.
+                    validate_timestamp_sig(sig_alg, hash_alg, sig_val, &tbs, &signing_key_der)?;
+                } else {
+                    // NOTE: We're keeping the WASM-specific async path alive for now because it
+                    // supports more signature algorithms. Look for future WASM platform to provide
+                    // the opportunity to unify.
+                    validate_timestamp_sig_async(
```

Comment:
Does validation need an async call path?

Reply:
It has an async path for wasm.